packages/react/src/components/DetailsList/DetailsHeader.base.tsx

```diff
@@ -383,17 +386,32 @@ export class DetailsHeaderBase
   };
 
   private _onDrop = (item?: any, event?: DragEvent): void => {
+    const { keyboardColumnReorderProps } = this.props;
+
     // Safe to assume this is defined since we're handling a drop event
     const columnReorderProps = this._getColumnReorderProps()!;
-
     // Target index will not get changed if draggeditem is after target item.
-    if (this._draggedColumnIndex >= 0 && event) {
-      const targetIndex =
-        this._draggedColumnIndex > this._currentDropHintIndex
-          ? this._currentDropHintIndex
-          : this._currentDropHintIndex - 1;
-      const isValidDrop = this._isValidCurrentDropHintIndex();
-      event.stopPropagation();
+    if ((this._draggedColumnIndex >= 0 && event) || keyboardColumnReorderProps) {
+      let targetIndex;
+      if (keyboardColumnReorderProps) {
+        targetIndex = this._isCheckboxColumnHidden()
```

Comment:
I pulled this code to run the story locally and I think that this reordering logic is not working correctly. Repro steps:

1. Load the DetailsList keyboard accessibility story
2. Set the left frozen columns to 0
3. Click the "key" header and select "reorder"
4. Enter "1" in the dialog and click "reorder"

**Expected behavior**

The "key" column should be the first (left most) column.

**Actual behavior**

The "key" column does not move

Similarly, if I try to reorder the "key" column to be the fourth column it ends up being the third column.

https://user-images.githubusercontent.com/93940821/143071308-c6a22f1a-a67f-4b54-9915-6aaf757f48dd.mp4

Reply:
DetailsList considers the select/select all checkboxes as a column so that's technically the first column in this example. It's also frozen by default which is why nothing happened when you tried to move the column `key` to index 1. When you tried to move `key` to index 4, it was technically moved to the correct column but I agree that it's a bit confusing so I'll modify the example to offset the inputted index in a reorder by 1 😃

Reply:
changes made here: https://github.com/microsoft/fluentui/pull/20679/commits/8e2685820f74e15a96d6061e23fc07b66b2ece3d